### ISSUE TITLE
Make it easier to transition between versions

### ISSUE THREAD

**Issue by rylev** [rust-lang/measureme#178]
Currently the process of moving from one version of measureme, summarize, and others to a newer version can be error prone. Generally newer versions of the tools emit and read different versions of the on-disk `.mm_profdata` file format. 

When reading a the profile data file, the parser first checks that the version of the file is exactly equal to the version the tool expects. https://github.com/rust-lang/measureme/blob/b9cccd7ad4c859a5e0e3dd6bff3daac7a190bdd7/measureme/src/file_header.rs#L77

If it's not equal the parser returns an error. 

In the end, we ideally want so that users rarely encounter errors, and when they do they instructions for fixing the error are simple and reliable. 

There are several improvements that we can do to address this issue.

## Better error messages

The most lightweight improvement would be to improve error messages when the tool encounters a file format version that it does not expect. 

If we detect the file version is less than the current version, tell users that they should regenerate the file if possible or install an older version of the tool that supports that file version. We could optionally, keep track of which versions of measure and co. work with which versions of the file format, and report to the user which one they should use. 

If we detect the file version is greater than the latest version, we can encourage users to upgrade the tools to the latest version. 

## Make the parser understand older file versions

Requiring the parser to understand at least the previous version of the file format will make the scenario where the user upgrades or installs the newest version of the measureme tool suite, but rustc has not yet been updated to emit the new file version. 

When changes to the file format are relatively small, this should be fairly trivial to do, but it may prove difficult if we ever need to dramatically change the file format. 

This does not help in the situation where the tool is older than the file format which can happen when users go longer periods between using the measureme tool suite. 

## Other considerations

* [**meta-executable**](https://github.com/rust-lang/measureme/issues/164): depending on how implement a meta-executable for the measureme tool suite, we could consider making it version aware and have it "fallback" to older versions when parsing fails. 
* perf.rlo [currently tries](https://github.com/rust-lang/rustc-perf/blob/d7f85aa72b2c648233e8513f1af85e697e32e726/collector/src/rustc-fake.rs#L345-L367) to use the `summarize` binary on its path (which is 9.1.2 currently) and when it fails it falls back version 9.0. If that then fails, everything fails. This hack works for now, but will require updating as future releases of measureme come out. Ideally perf.rlo shouldn't need to worry about this.

**Comment by Mark-Simulacrum** [comment]
Note that perf.rlo also uses the library to generate .crox and similar links on the frontend. Currently, that's tied to just one version, which means currently 0.8 and older version generated files don't work.

Ideally, I think, measureme itself would multiplex between at least 2 last versions. I imagine in theory this isn't too hard, and we could gate it on a feature flag so e.g. rustc doesn't need to worry about extra bloat.

**Comment by wesleywiser** [comment]
We've already split the reading and writing parts of the crate in half: `measureme` contains just the code to write the current version of the format (and is what `rustc` depends on) while `analyzeme` implements reading and analyzing the format. 

I'm in favor of making `analyzeme` support reading & processing the last 2-3 versions of the file format.

**Comment by michaelwoerister** [comment]
OK, I have a prototype at https://github.com/michaelwoerister/measureme/tree/support-multiple-file-formats2 but it is not pretty `:/`

The `analyzeme` uses many types from `measureme` (e.g. `RawEvent`, `StringId`, etc) -- but we cannot use these types directly for decoders of older file formats. E.g. the current `RawEvent` in `measureme` might be different from the `RawEvent` in a previous version, so if the decoder references the current version, it might assume a wrong data layout. I "solved" this by creating a "snapshot" crate that re-exports a specific version of `measureme` and let the decoder for the previous file format reference the types in there. That's not very pretty but at least we don't have to manually copy the type definitions out of `measureme`. 

It turns out, however, that with this approach we cannot just re-use the implementation of e.g. `StringTable` even though it effectively has not changed. @rylev and I initially thought we could just re-export `v7::StringTable` as `v8::StringTable` in such a case. But that doesn't work because `v8::StringTable` (the "current" implementation) references `StringId` from `measureme` while `v7::StringTable` references `StringId` from `mm_snapshot_9_1_2`. So I essentially had to copy-and-paste the `StringTable` implementation from `v7` into `v8`. 

I'm not sure if this approach is viable at all.

**Comment by Mark-Simulacrum** [comment]
I think the approach I took when trying this out (6 months or something ago) was to try and "convert" from the old types to the latest types, which seemed to work decently well? That approach likely requires that the conversion code live in the same crate/module so it can have private field access and such.

Alternatively, even if we cannot get programmatic access to work across the different versions, it would be nice to have a summarize binary at least (which is the only one used on perf.r-l.o's collector) that multiplexes internally amongst the different versions. In theory that shouldn't need any conversion between types, right? Since the output format is presumably so far pretty stable, or at least additive.

Programmatic access if necessary can likely be multiplexed in perf.r-l.o's codebase directly, rather than here, to limit the amount of work needed. (Since only a subset of the APIs needs multiplexing then).

**Comment by rylev** [comment]
For easier access, here's a diff view of the [changes](https://github.com/rust-lang/measureme/compare/master...michaelwoerister:support-multiple-file-formats2). 

@michaelwoerister discussed a different approach which would have the old implementations of the file format copied into `analyzeme`. This could work, but might be too much of a maintenance burden and/or too error prone.

**Comment by michaelwoerister** [comment]
After discussing with @rylev, this is a much better approach (at least in my opinion): 
https://github.com/michaelwoerister/measureme/tree/support-multiple-file-formats3

Here we make `analyzeme 10.0.0` just depend on `analyzeme 9.1.2` in its Cargo.toml. This way it is pretty easy to safely use the old implementation of everything and just converting things right at the end.

**Comment by michaelwoerister** [comment]
Convenience link: https://github.com/rust-lang/measureme/compare/master...michaelwoerister:support-multiple-file-formats3

**Comment by michaelwoerister** [comment]
One problem I noticed with making the crate depend on itself is that we would be creating a linked list of crate versions that cannot be shortened again. Even if the newest version only directly depends on the last two versions,  those previous versions would in turn keep depending on even older versions:
```
v_9_1_2 <--- v_10_0_0 <--- v_11_0_0 <--- v_12_0_0 <--- v_13_0_0
   ^             ^           |  ^            |             |
   |             |           |  |            |             |
   +-------------------------+  |            |             |
                 |              |            |             |
                 +---------------------------+             |
                                |                          |
                                +--------------------------+
```
e.g. v13 only depends on v12 and v11 directly, but v11 would keep depending on v10 and v9_1_2 and so on. The crate graph would keep growing and growing with no way of ever dropping old versions from it. So I propose that we introduce a decoder crate (called `decodeme` of course `:)`) that defines `Event` and `LightweightEvent` and acts as an intermediary between `measureme` and `analyzeme`:

```
mm_9_1_2  <--- decoder_9_1_2  <----+
                                   |
mm_10_0_0 <--- decoder_10_0_0 <----+---- analyzeme_10_0_0

[...]

mm_15_0_0 <--- decoder_15_0_0 <----+
                                   |
mm_16_0_0 <--- decoder_16_0_0 <----+
                                   |
mm_17_0_0 <--- decoder_17_0_0 <----+---- analyzeme_17_0_0
```
That way we can bound the number of old crate versions we pull in. Tools would keep depending on only the newest version of `analyzeme`. It will need a little bit of shuffling things around internally but I think it is worth the effort.

Alternatively we could move the parts that would end up in `decodeme` into `measureme`. That would simplify things at the expensive of having more stuff in `measureme` that is not needed by the compiler.

**Comment by michaelwoerister** [comment]
Here is a version of v9.1.2 with encoding-specific things from `analyzeme` factored out into a `decodeme` crate. It turned out to be pretty clean, in my opinion: https://github.com/rust-lang/measureme/compare/9.1.2...michaelwoerister:introduce-decodeme-to-v9

**Comment by michaelwoerister** [comment]
If we decide to go ahead with this approach my plan would be to:
- Add this branch to the upstream repo and tag it as `9.1.3` (but not necessarily create an actual crates.io release for it)
- Do a similar refactoring for the upstream/main branch. This would be the last time we have to duplicate work for `decodeme`.
- Add support for v9 encodings by turning `EventDecoder` in `ProfilingData` into a trait object that gets initialized depending on the encoding version in the file header to either be an `EventDecoder` from `decodeme 9.1.3` or `decodeme 10.0.0`.

How does that sound?
